subsys/shell/backends/shell_bt.c

```diff
@@ -0,0 +1,282 @@
+{
+	shell_bt_st.bt.notif_enabled = !(value == 0);
+}
+
+BT_GATT_SERVICE_DEFINE(shell_svc,
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_SHELL_SERVICE),
+	BT_GATT_CHARACTERISTIC(BT_UUID_SHELL_CHAR,
+		BT_GATT_CHRC_WRITE_WITHOUT_RESP |
+			BT_GATT_CHRC_NOTIFY,
+			BT_GATT_PERM_WRITE,
+		NULL, shell_bt_chr_write, NULL),
+	BT_GATT_CCC(shell_ccc_cfg_changed,
+		BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+);
+
+static int bt_shell_notify(struct bt_conn *conn, const uint8_t *data, size_t len)
+{
+	/** @NOTE: Split message in chunks accomodating for current GATT MTU */
+	uint32_t max_len = bt_gatt_get_mtu(conn) - 3;
+	int err = -1;
+	size_t bytes_transferred = 0;
+	uint32_t length = 0;
+
+	do {
+		length = MIN(max_len, (len - bytes_transferred));
+
+		if (length > 0) {
+			LOG_DBG("tx-len: %d, mtu: %d", length, max_len);
+			err = bt_gatt_notify(conn, &shell_svc.attrs[1],
+				&data[bytes_transferred], length);
+			if (err) {
+				(void)bt_conn_disconnect(conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+				return err;
+			}
+			bytes_transferred += length;
+		}
+	} while (length > 0);
+
+	return len;
+}
+
+/**
+ * @brief Get a temporary copy of the current BT connection object
+ *
+ * @details The motivation behind this helper function is to safely use the
+ * connection object to send notifications in the write API. Since it is not
+ * safe to hold the mutex while calling BT APIs, we are relying on this
+ * reference copy to safely use the connection object.
+ *
+ * @note The caller needs to free this additional connection reference through
+ * @ref bt_conn_unref.
+ *
+ * @retval Connection Reference.
+ */
+static struct bt_conn *get_conn_ref_copy(void)
+{
+	struct bt_conn *conn_ref = NULL;
+
+	(void)k_mutex_lock(shell_bt_st.bt.conn.mutex, K_FOREVER);
+
+	if (shell_bt_st.bt.conn.conn) {
+		conn_ref = bt_conn_ref(shell_bt_st.bt.conn.conn);
+	}
+
+	(void)k_mutex_unlock(shell_bt_st.bt.conn.mutex);
+
+	return conn_ref;
+}
+
+static int init(const struct shell_transport *transport, const void *config,
+		shell_transport_handler_t evt_handler, void *context)
+{
+	ARG_UNUSED(config);
+
+	struct shell_bt *sh_bt = (struct shell_bt *)transport->ctx;
+
+	sh_bt->shell.context = context;
+	sh_bt->shell.handler = evt_handler;
+
+	return 0;
+}
+
+static int uninit(const struct shell_transport *transport)
+{
+	ARG_UNUSED(transport);
+
+	return 0;
+}
+
+static int enable(const struct shell_transport *transport, bool blocking_tx)
+{
+	ARG_UNUSED(transport);
+	ARG_UNUSED(blocking_tx);
+
+	return 0;
+}
+
+static int write(const struct shell_transport *transport,
+		const void *data, size_t length, size_t *cnt)
+{
+	int err = 0;
+	struct bt_conn *conn = NULL;
+
+	*cnt = length;
+
+	LOG_DBG("%s: %d", __func__, length);
+	LOG_HEXDUMP_DBG(data, length, "data_write");
+
+	/* It is not safe to use BT APIs with the conn mutex locked: rely on a reference copy */
+	(void)k_mutex_lock(shell_bt_st.bt.conn.mutex, K_FOREVER);
```

Comment:
The mutex is taken inside `get_conn_ref_copy`, so it should not be taken here as well.

Side note: In fact, taking the mutex twice like this works only because the Zephyr mutex is 're-entrant'. Not all 'mutex' implementations are re-entrant. It it contested in the general CS community whether mutex re-entrancy is a desirable feature at all. I think it's awkward and more likely to hide bugs than be useful. :) /purist

Reply:
Gotchya - Removed redundant mutex guard.

Let's keep us all happy if we can :smile: